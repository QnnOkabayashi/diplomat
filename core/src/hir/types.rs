//! Types that can be exposed in Diplomat APIs.

use super::{paths, PrimitiveType, TypeContext, TypeLifetime};

/// Type that may be used as an output.
pub enum ReturnableType {
    Type(Type),
    OutType(OutType),
}

/// Type that can only be used as an output.
pub enum OutType {
    Primitive(PrimitiveType),
    Opaque(paths::OutOpaque),
    Struct(paths::ReturnableStruct),
    Enum(paths::Enum),
    Slice(Slice),
}

/// Type that may be used as input or output.
pub enum Type {
    Primitive(PrimitiveType),
    Opaque(paths::Opaque),
    Struct(paths::Struct),
    Enum(paths::Enum),
    Slice(Slice),
}

/// Type that can appear in the `self` position.
pub enum SelfType {
    Opaque(paths::SelfOpaque),
    Struct(paths::Struct),
    Enum(paths::Enum),
}

#[derive(Copy, Clone)]
pub enum Slice {
    /// A string slice, e.g. `&str`.
    Str(TypeLifetime),

    /// A primitive slice, e.g. `&mut [u8]`.
    Primitive(Ref, PrimitiveType),
}

#[derive(Copy, Clone)]
pub enum Mutability {
    Mutable,
    Immutable,
}

#[derive(Copy, Clone)]
pub struct Ref {
    pub lifetime: TypeLifetime,
    pub mutability: Mutability,
}

impl Type {
    /// Return the number of fields and leaves that will show up in the [`LifetimeTree`]
    /// returned by [`Param::lifetime_tree`] and [`ParamSelf::lifetime_tree`].
    ///
    /// This method is used to calculate how much space to allocate upfront.
    pub(super) fn field_leaf_lifetime_counts(&self, tcx: &TypeContext) -> (usize, usize) {
        match self {
            Type::Struct(ty) => ty.resolve(tcx).fields.iter().fold((1, 0), |acc, field| {
                let inner = field.ty.field_leaf_lifetime_counts(tcx);
                (acc.0 + inner.0, acc.1 + inner.1)
            }),
            Type::Opaque(_) | Type::Slice(_) => (0, 1),
            Type::Primitive(_) | Type::Enum(_) => (0, 0),
        }
    }
}

impl Slice {
    /// Returns the [`TypeLifetime`] contained in either the `Str` or `Primitive`
    /// variant.
    pub fn lifetime(&self) -> &TypeLifetime {
        match self {
            Slice::Str(lifetime) => lifetime,
            Slice::Primitive(reference, _) => &reference.lifetime,
        }
    }
}
