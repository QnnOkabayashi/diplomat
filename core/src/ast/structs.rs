use serde::{Deserialize, Serialize};

use super::docs::Docs;
use super::{Ident, LifetimeEnv, Method, NamedLifetime, PathType, TypeName};

/// A struct declaration in an FFI module that is not opaque.
#[derive(Clone, PartialEq, Eq, Hash, Serialize, Deserialize, Debug)]
pub struct Struct {
    pub name: Ident,
    pub docs: Docs,
    pub lifetimes: LifetimeEnv,
    pub fields: Vec<(Ident, TypeName, Docs)>,
    pub methods: Vec<Method>,
}

impl Struct {
    // WARNING: this could probably be made more optimized, but here's what I'm thinking:
    // Take a predicate which accepts a `usize` denoting the position the lifetime
    // occurred at in the struct definition, and return true if that lifetime
    // should be kept. By working with indices instead of lifetime names, we're
    // robust against renaming lifetimes in different blocks.
    pub fn borrowed_fields<'a, F>(&'a self, mut pred: F) -> Vec<&(Ident, TypeName, Docs)>
    where
        F: FnMut(usize) -> bool,
    {
        // A mapping from def indices to env indices
        // The index of each element corresponds to the index in the `syn::LifetimeDef`
        // that it came from
        // The value at each index corresponds to where that particular lifetime
        // lives in the `LifetimeEnv`.
        // This would have to be stored in `Self` before hand.
        // let def_idx_to_env_idx: Vec<usize> = vec![];

        // Since lifetime names can be totally different, we care about the indices,
        // NOT the names.
        // let _lts_we_care_about: Vec<&NamedLifetime> = def_idx_to_env_idx
        //     .iter()
        //     .enumerate()
        //     .filter_map(|(idx, ptr)| {
        //         if pred(idx) {
        //             // pretend this works and is safe
        //             Some(&self.lifetimes.nodes[ptr].lifetime)
        //         } else {
        //             None
        //         }
        //     })
        //     .collect::<Vec<_>>();

        // One strategy we could do to assign indices to lifetimes: Use the
        // index of the lifetime in the `LifetimeEnv`, NOT the index at which
        // it appears in the struct definition.
        // This requires remembering the order that lifetimes are read in
        todo!()
    }
}

impl From<&syn::ItemStruct> for Struct {
    /// Extract a [`Struct`] metadata value from an AST node.
    fn from(strct: &syn::ItemStruct) -> Struct {
        let self_path_type = PathType::extract_self_type(strct);

        Struct {
            name: (&strct.ident).into(),
            docs: Docs::from_attrs(&strct.attrs),
            lifetimes: LifetimeEnv::from(&strct.generics),
            fields: strct
                .fields
                .iter()
                .map(|field| {
                    // Non-opaque tuple structs will never be allowed
                    let name = field
                        .ident
                        .as_ref()
                        .map(Into::into)
                        .expect("non-opaque tuples structs are disallowed");
                    let type_name = TypeName::from_syn(&field.ty, Some(self_path_type.clone()));
                    let docs = Docs::from_attrs(&field.attrs);

                    (name, type_name, docs)
                })
                .collect(),
            methods: vec![],
        }
    }
}

/// A struct annotated with [`diplomat::opaque`] whose fields are not visible.
/// Opaque structs cannot be passed by-value across the FFI boundary, so they
/// must be boxed or passed as references.
#[derive(Clone, Serialize, Deserialize, Debug, Hash, PartialEq, Eq)]
pub struct OpaqueStruct {
    pub name: Ident,
    pub docs: Docs,
    pub lifetimes: LifetimeEnv,
    pub methods: Vec<Method>,
}

impl From<&syn::ItemStruct> for OpaqueStruct {
    /// Extract a [`OpaqueStruct`] metadata value from an AST node.
    fn from(strct: &syn::ItemStruct) -> OpaqueStruct {
        OpaqueStruct {
            name: (&strct.ident).into(),
            docs: Docs::from_attrs(&strct.attrs),
            lifetimes: LifetimeEnv::from(&strct.generics),
            methods: vec![],
        }
    }
}

#[cfg(test)]
mod tests {
    use insta::{self, Settings};

    use syn;

    use super::Struct;

    #[test]
    fn simple_struct() {
        let mut settings = Settings::new();
        settings.set_sort_maps(true);

        settings.bind(|| {
            insta::assert_yaml_snapshot!(Struct::from(&syn::parse_quote! {
                /// Some docs.
                #[diplomat::rust_link(foo::Bar, Struct)]
                struct MyLocalStruct {
                    a: i32,
                    b: Box<MyLocalStruct>
                }
            }));
        });
    }
}
