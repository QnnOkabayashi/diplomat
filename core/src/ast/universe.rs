
use std::collections::BTreeMap;

use crate::Env;

use super::{Docs, Ident, LifetimeEnv, Method, PrimitiveType, Lifetime, Mutability, Path, ModSymbol, CustomType, OpaqueStruct, Enum, TypeName};

pub struct Environment {
    inputs: BTreeMap<InputPath, InputStruct>,
    outputs: BTreeMap<OutputPath, OutputStruct>,
    input_outputs: BTreeMap<InputOutputPath, InputOutputStruct>,
    opaques: BTreeMap<OpaquePath, OpaqueStruct>,
    enums: BTreeMap<EnumPath, Enum>,
}

impl Environment {
    pub fn from_env(env: &Env) -> Self {
        let mut table = BTreeMap::new();
        // first read in all the paths
        // then create a table with each path and a bool for "input" and "output"
        for (path, _, symbol) in env.iter_items() {
            if let ModSymbol::CustomType(custom) = symbol {
                match custom {
                    CustomType::Struct(_) => {
                        table.insert(path, (false, false));
                    }
                    CustomType::Opaque(_) => todo!(),
                    CustomType::Enum(_) => todo!(),
                }
            }
        }

        for (path, ident, symbol) in env.iter_items() {
            if let ModSymbol::CustomType(custom) = symbol {
                for method in custom.methods() {
                    if method.self_param.is_some() {
                        table.get_mut(path).unwrap().0 = true;
                    }
                    for input in method.params.iter() {
                        if let TypeName::Named(ref named) = input.ty {
                            table.get_mut(named.path()).unwrap().0 = true;
                        }
                        // mark type recursively
                    }


                    if let Some(ref return_type) = method.return_type {
                        if let TypeName::Named(ref named) = return_type {
                            table.get_mut(named.path()).unwrap().1 = true;
                        }
                        // mark type recursively
                    }
                }
            }
            // go through all the inputs of each method recursively, marking each 
            //   custom type as input=true
            // similar story for output types and output=true
            // if there are types that aren't input or output types, they're never used so complain
            // Then look at each struct's fields (nonrecursively) and verify that all other
            // fields that aren't recursive also fit criteria.
            // if some don't fit (e.g. owned ptr on input type), complain and and show a nice
            // error message.
        }

        let mut inputs = BTreeMap::new();
        let mut outputs = BTreeMap::new();
        let mut input_outputs = BTreeMap::new();
        for (path, (is_input, is_output)) in table {
            match (is_input, is_output) {
                (true, true) => input_outputs.insert(InputOutputPath(path.clone(), todo!())),
                (true, false) => inputs.insert(InputPath(path.clone()), todo!()),
                (false, true) => outputs.insert(OutputPath(path.clone()), todo!()),
                (false, false) => panic!("unused type"),
            };
        }

        Environment {
            inputs,
            outputs,
            input_outputs,
            opaques: todo!(),
            enums: todo!(),
        }
    }

    fn input_type(&self, path: &InputPath) -> &InputStruct {
        self.inputs.get(path).unwrap()
    }

    fn output_type(&self, path: &OutputPath) -> &OutputStruct {
        self.outputs.get(path).unwrap()
    }

    fn input_output_type(&self, path: &InputOutputPath) -> &InputOutputStruct {
        self.input_outputs.get(path).unwrap()
    }

    fn opaque_type(&self, path: &OpaquePath) -> &OpaqueStruct {
        self.opaques.get(path).unwrap()
    }

    fn enum_type(&self, path: &EnumPath) -> &Enum {
        self.enums.get(path).unwrap()
    }
}



#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct InputPath(Path);


#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct OutputPath(Path);


#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct InputOutputPath(Path);


#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct OpaquePath(Path);


#[derive(Debug, Clone, PartialEq, Eq, PartialOrd, Ord)]
pub struct EnumPath(Path);

/// A non-opaque struct that can be an input to a function.
pub struct InputStruct {
    pub name: Ident,
    pub docs: Docs,
    pub lifetimes: LifetimeEnv,
    pub fields: Vec<(Ident, InputType, Docs)>,
    pub methods: Vec<Method>,
}

/// A non-opaque struct that can be an output to a function.
pub struct OutputStruct {
    pub name: Ident,
    pub docs: Docs,
    pub lifetimes: LifetimeEnv,
    pub fields: Vec<(Ident, OutputType, Docs)>,
    pub methods: Vec<Method>,
}

/// A non-opaque struct that can be both an input or an output to a function.
pub struct InputOutputStruct {
    pub name: Ident,
    pub docs: Docs,
    pub lifetimes: LifetimeEnv,
    pub fields: Vec<(Ident, InputOutputType, Docs)>,
    pub methods: Vec<Method>,
}

pub enum InputPathKind {
    Input(InputPath),
    InputOutput(InputOutputPath),
}

pub enum OutputPathKind {
    Output(OutputPath),
    InputOutput(InputOutputPath),
}

pub enum InputType {
    Primitive(PrimitiveType),
    Opaque(Nullability, OpaquePath), // always borrowed
    Struct(InputPathKind),
    Enum(EnumPath),
    Slice(Slice),
    Writable,
    Unit,
}

pub enum OutputType {
    Primitive(PrimitiveType),
    Opaque(Nullability, Pointer, OpaquePath),
    Struct(OutputPathKind),
    Enum(EnumPath),
    Slice(Slice),
    Result(Box<OutputType>, Box<OutputType>),
    Unit,
}

pub enum InputOutputType {
    Primitive(PrimitiveType),
    Opaque(Nullability, OpaquePath),
    Struct(InputOutputStruct),
    Enum(EnumPath),
    Slice(Slice),
    Unit,
}

pub enum Slice {
    Str(Lifetime),
    Primitive(Lifetime, Mutability, PrimitiveType),
}

pub enum Nullability {
    Nullable,
    NonNull,
}

pub enum Pointer {
    Boxed,
    Reference(Lifetime),
}

