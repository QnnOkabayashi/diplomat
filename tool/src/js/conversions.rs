use diplomat_core::Env;
use std::fmt;
use std::fmt::Write;

use diplomat_core::ast;

use super::display;
use super::types::{return_type_form, ReturnTypeForm};
use crate::layout;

/// TODO: docs
#[allow(clippy::ptr_arg)] // false positive, rust-clippy#8463, fixed in 1.61
pub fn gen_value_js_to_rust(
    param_name: &ast::Ident,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    env: &Env,
    borrowed_lifetimes: &[&ast::NamedLifetime],
    pre_logic: &mut Vec<String>,
    invocation_params: &mut Vec<String>,
    post_logic: &mut Vec<String>,
) {
    match typ {
        ast::TypeName::StrReference(lifetime) | ast::TypeName::PrimitiveSlice(lifetime, ..) => {
            let js_param_name = if let ast::TypeName::PrimitiveSlice(.., prim) = typ {
                let js_param_name = ast::Ident::from(format!("{}_diplomat_slice", param_name));
                pre_logic.push(format!(
                    "let {js_param_name} = diplomatRuntime.DiplomatBuf.slice(wasm, {param_name}, {align});",
                    align = layout::primitive_size_alignment(*prim).align()
                ));
                js_param_name
            } else {
                let js_param_name = ast::Ident::from(format!("{}_diplomat_str", param_name));
                pre_logic.push(format!(
                    "let {js_param_name} = diplomatRuntime.DiplomatBuf.str(wasm, {param_name});"
                ));
                js_param_name
            };

            invocation_params.push(format!("{js_param_name}.ptr"));
            invocation_params.push(format!("{js_param_name}.size"));

            match lifetime {
                ast::Lifetime::Named(ref named) if borrowed_lifetimes.contains(&named) => {
                    // TODO: don't create an edge if this is a str/slice,
                    // since they actually copy and thus don't rely on the buffer.
                    post_logic.push(format!(
                        "diplomat_out.__{param_name}_lifetime_guard = {js_param_name};",
                    ));
                }
                _ => {
                    post_logic.push(format!("{js_param_name}.free();"));
                }
            }
        }
        ast::TypeName::Primitive(ast::PrimitiveType::char) => {
            // we use the spread operator here to count codepoints
            // codePointAt() does not return surrogate pairs if there are multiple
            invocation_params.push(format!(
                "diplomatRuntime.extractCodePoint({p}, '{p}')",
                p = param_name
            ));
        }
        ast::TypeName::Box(_) => {
            invocation_params.push(format!("{}.underlying", param_name));
        }
        ast::TypeName::Reference(lifetime, _mut, _typ) => {
            match lifetime {
                ast::Lifetime::Named(named) if borrowed_lifetimes.contains(&named) => {
                    // We want to hold the reference here
                }
                _ => {}
            }
            invocation_params.push(format!("{param_name}.underlying"));
        }
        ast::TypeName::Named(path_type) => {
            match path_type.resolve(in_path, env) {
                ast::CustomType::Struct(struct_type) => {
                    // == THE GAMEPLAN ==
                    // do `Struct::borrowed_lifetimes`, which takes
                    // a filter that tells which indices of lifetimes we need.
                    // Let that filter take `i`, an index. do `path_type.lifetimes[i]`
                    // to get what the lifetime is in the current `LifetimeEnv`
                    // (not the `Struct`'s env!), and then see if that lifetime
                    // in `borrowed_lifetimes`. If it is,
                    // have the filter return `true`, otherwise `false`.
                    // The `Struct::borrowed_lifetimes` method will then return
                    // a list of all the relevant lifetimes in the `Struct`'s
                    // `LifetimeEnv` scope, which we can pass down to further
                    // recursive calls.
                    // Once we get this all working, we can start thinking about indices.
                    // ==================

                    // Shadow the borrowed lifetime from this invocation so they're
                    // ready for the recursive invocation.
                    let borrowed_lifetimes = struct_type.borrowed_lifetimes(|i| {
                        match path_type.lifetimes.get(i) {
                            Some(ast::Lifetime::Static) => false,
                            Some(ast::Lifetime::Named(ref named)) => {
                                borrowed_lifetimes.contains(&named)
                            }
                            Some(ast::Lifetime::Anonymous) | None => {
                                panic!("lifetime elision is unsupported, {} missing lifetime at index {}", path_type, i)
                            }
                        }
                    });

                    for (field_name, field_type, _) in struct_type.fields.iter() {
                        let field_extracted_name =
                            format!("diplomat_{}_extracted_{}", struct_type.name, field_name)
                                .into();
                        pre_logic.push(format!(
                            "const {} = {}[\"{}\"];",
                            field_extracted_name, param_name, field_name
                        ));

                        gen_value_js_to_rust(
                            &field_extracted_name,
                            field_type,
                            in_path,
                            env,
                            &borrowed_lifetimes[..],
                            pre_logic,
                            invocation_params,
                            post_logic,
                        );
                    }
                }
                ast::CustomType::Enum(enm) => {
                    invocation_params.push(format!("{}_js_to_rust[{}]", enm.name, param_name));
                }

                ast::CustomType::Opaque(_) => {
                    panic!("Opaque types cannot be sent as values");
                }
            }
        }
        _ => invocation_params.push(param_name.to_string()),
    }
}

/// TODO: docs
pub fn gen_value_rust_to_js<W: fmt::Write>(
    value_expr: &str,
    typ: &ast::TypeName,
    in_path: &ast::Path,
    borrowed_params: &ast::BorrowedParams,
    env: &Env,
    out: &mut W,
) -> fmt::Result {
    match typ {
        ast::TypeName::Named(path_type) => match path_type.resolve(in_path, env) {
            ast::CustomType::Struct(strct) => {
                let strct_size_align = layout::type_size_alignment(typ, in_path, env);
                let needs_buffer = return_type_form(typ, in_path, env);
                if needs_buffer != ReturnTypeForm::Complex {
                    todo!("Receiving structs that don't need a buffer: {}", strct.name)
                    // recursively call on single field
                }
                write!(
                    out,
                    "{}",
                    display::iife(|mut f| {
                        writeln!(
                            f,
                            "const diplomat_receive_buffer = wasm.diplomat_alloc({}, {});",
                            strct_size_align.size(),
                            strct_size_align.align(),
                        )?;
                        writeln!(f, "{};", value_expr)?;
                        writeln!(
                            f,
                            "const out = new {}(diplomat_receive_buffer);",
                            strct.name
                        )?;

                        writeln!(
                            f,
                            "wasm.diplomat_free(diplomat_receive_buffer, {size}, {align});",
                            size = strct_size_align.size(),
                            align = strct_size_align.align()
                        )?;

                        writeln!(f, "return out;")
                    })
                )?;
            }

            ast::CustomType::Enum(enm) => {
                write!(out, "{}_rust_to_js[{}]", enm.name, value_expr)?;
            }

            ast::CustomType::Opaque(_) => {
                panic!("Opaque types cannot be used in value position")
            }
        },

        ast::TypeName::Box(underlying) => {
            gen_rust_reference_to_js(
                underlying.as_ref(),
                in_path,
                value_expr,
                borrowed_params,
                env,
                out,
            )?;
        }

        ast::TypeName::Option(_) => {
            gen_rust_reference_to_js(typ, in_path, value_expr, borrowed_params, env, out)?;
        }

        ast::TypeName::Result(ok, err) => {
            let (ok_offset, result_size_align) =
                layout::result_ok_offset_size_align(ok, err, in_path, env);
            let needs_buffer = return_type_form(typ, in_path, env) == ReturnTypeForm::Complex;
            write!(
                out,
                "{}",
                display::iife(|mut f| {
                    let (size, align) = (result_size_align.size(), result_size_align.align());
                    if needs_buffer {
                        writeln!(
                            f,
                            "const diplomat_receive_buffer = wasm.diplomat_alloc({size}, {align});"
                        )?;
                        writeln!(f, "{};", value_expr)?;
                        writeln!(
                            f,
                            "const is_ok = {};",
                            display::expr(|mut f| {
                                gen_rust_reference_to_js(
                                    &ast::TypeName::Primitive(ast::PrimitiveType::bool),
                                    in_path,
                                    &format!("diplomat_receive_buffer + {}", ok_offset),
                                    borrowed_params,
                                    env,
                                    &mut f,
                                )
                            })
                        )?;
                        writeln!(
                            f,
                            "if (is_ok) {is_true} else {is_false}",
                            is_true = display::block(|mut f| {
                                writeln!(
                                    f,
                                    "const ok_value = {};",
                                    display::expr(|mut f| {
                                        gen_rust_reference_to_js(
                                            ok.as_ref(),
                                            in_path,
                                            "diplomat_receive_buffer",
                                            borrowed_params,
                                            env,
                                            &mut f,
                                        )
                                    })
                                )?;
                                writeln!(
                                    f,
                                    "wasm.diplomat_free(diplomat_receive_buffer, {size}, {align});"
                                )?;
                                writeln!(f, "return ok_value;")
                            }),
                            is_false = display::block(|mut f| {
                                writeln!(
                                    f,
                                    "const throw_value = {};",
                                    display::expr(|mut f| {
                                        gen_rust_reference_to_js(
                                            err.as_ref(),
                                            in_path,
                                            "diplomat_receive_buffer",
                                            borrowed_params,
                                            env,
                                            &mut f,
                                        )
                                    })
                                )?;
                                writeln!(
                                    f,
                                    "wasm.diplomat_free(diplomat_receive_buffer, {size}, {align});"
                                )?;
                                writeln!(f, "throw new diplomatRuntime.FFIError(throw_value);")
                            })
                        )
                    } else {
                        writeln!(f, "const is_ok = {} == 1;", value_expr)?;
                        writeln!(
                            f,
                            "if (!is_ok) {}",
                            display::block(|mut f| {
                                writeln!(f, "throw new diplomatRuntime.FFIError({{}});")
                            })
                        )
                    }
                })
            )?;
        }

        ast::TypeName::Primitive(_prim) => {
            // TODO(#63): wrap with appropriate types for large widths
            write!(out, "{}", value_expr)?;
        }

        ast::TypeName::Reference(.., underlying) => {
            gen_rust_reference_to_js(
                underlying.as_ref(),
                in_path,
                value_expr,
                borrowed_params,
                env,
                out,
            )?;
        }
        ast::TypeName::Writeable => todo!(),
        ast::TypeName::StrReference(..) => {
            gen_rust_reference_to_js(typ, in_path, value_expr, borrowed_params, env, out)?;
        }
        ast::TypeName::PrimitiveSlice(..) => todo!(),
        ast::TypeName::Unit => write!(out, "{}", value_expr)?,
    }

    Ok(())
}

/// TODO: docs
pub fn gen_rust_reference_to_js<W: fmt::Write>(
    underlying: &ast::TypeName,
    in_path: &ast::Path,
    value_expr: &str,
    borrowed_params: &ast::BorrowedParams,
    env: &Env,
    out: &mut W,
) -> fmt::Result {
    match underlying {
        ast::TypeName::Box(typ) | ast::TypeName::Reference(.., typ) => {
            gen_rust_reference_to_js(
                typ.as_ref(),
                in_path,
                &format!(
                    "(new Uint32Array(wasm.memory.buffer, {}, 1))[0]",
                    value_expr
                ),
                borrowed_params,
                env,
                out,
            )?;
        }
        ast::TypeName::Option(underlying) => match underlying.as_ref() {
            ast::TypeName::Box(..) | ast::TypeName::Reference(..) => write!(
                out,
                "{}",
                display::iife(|mut f| {
                    let ident = ast::Ident::from("option_ptr");
                    writeln!(
                        f,
                        "const {ident} = (new Uint32Array(wasm.memory.buffer, {}, 1))[0];",
                        value_expr
                    )?;
                    writeln!(
                        f,
                        "return ({ident} == 0) ? null : {};",
                        display::expr(|mut f| {
                            gen_value_rust_to_js(
                                ident.as_str(),
                                underlying.as_ref(),
                                in_path,
                                borrowed_params,
                                env,
                                &mut f,
                            )
                        })
                    )
                })
            )?,
            other @ (ast::TypeName::StrReference(..) | ast::TypeName::PrimitiveSlice(..)) => {
                panic!("`{}` is a fat pointer (ptr, length), and so it can't be stored behind an option.", other)
            }
            other => panic!(
                "`{0}` doesn't have the same alignment as ptr, so `Option<{0}>` is unsupported",
                other
            ),
        },

        ast::TypeName::Result(_, _) => {
            todo!("Receiving references to results")
        }

        ast::TypeName::Named(path_type) => {
            let custom_type = path_type.resolve(in_path, env);

            if let ast::CustomType::Enum(enm) = custom_type {
                write!(
                    out,
                    "{}_rust_to_js[{}]",
                    enm.name,
                    display::expr(|mut f| {
                        gen_rust_reference_to_js(
                            &ast::TypeName::Primitive(ast::PrimitiveType::isize),
                            in_path,
                            value_expr,
                            borrowed_params,
                            env,
                            &mut f,
                        )
                    })
                )?;
            } else {
                let params_borrowing_non_strrefs = borrowed_params
                    .borrowed_params
                    .iter()
                    .filter(|param| {
                        !param.ty.all_lifetimes(|lifetime, origin| {
                            matches!(
                                (lifetime, origin),
                                (ast::Lifetime::Named(_), ast::LifetimeOrigin::StrReference)
                            )
                        })
                    })
                    .collect::<Vec<_>>();

                if !borrowed_params.borrows_self && params_borrowing_non_strrefs.is_empty() {
                    write!(out, "new {}({})", custom_type.name(), value_expr)?;
                } else {
                    write!(
                        out,
                        "{}",
                        display::iife(|mut f| {
                            writeln!(f, "const out = new {}({});", custom_type.name(), value_expr)?;
                            if borrowed_params.borrows_self {
                                writeln!(f, "out.__this_lifetime_guard = this;")?;
                            }
                            for param in params_borrowing_non_strrefs.iter() {
                                writeln!(f, "out.__{0}_lifetime_guard = {0};", param.name)?;
                            }
                            writeln!(f, "return out;")
                        })
                    )?;
                }
            }
        }

        ast::TypeName::Primitive(prim) => {
            let prim_type = match prim {
                ast::PrimitiveType::i8 => "Int8Array",
                ast::PrimitiveType::u8 => "Uint8Array",
                ast::PrimitiveType::i16 => "Int16Array",
                ast::PrimitiveType::u16 => "Uint16Array",
                ast::PrimitiveType::i32 => "Int32Array",
                ast::PrimitiveType::u32 => "Uint32Array",
                ast::PrimitiveType::i64 => "BigInt64Array",
                ast::PrimitiveType::u64 => "BigUint64Array",
                ast::PrimitiveType::i128 => panic!("i128 not supported on JS"),
                ast::PrimitiveType::u128 => panic!("u128 not supported on JS"),
                ast::PrimitiveType::isize => "Int32Array",
                ast::PrimitiveType::usize => "Uint32Array",
                ast::PrimitiveType::f32 => "Float32Array",
                ast::PrimitiveType::f64 => "Float64Array",
                ast::PrimitiveType::bool => {
                    return write!(
                        out,
                        "(new Uint8Array(wasm.memory.buffer, {}, 1))[0] == 1",
                        value_expr
                    )
                }
                ast::PrimitiveType::char => {
                    return write!(
                        out,
                        "String.fromCharCode((new Uint32Array(wasm.memory.buffer, {}, 1))[0])",
                        value_expr
                    )
                }
            };

            write!(
                out,
                "(new {}(wasm.memory.buffer, {}, 1))[0]",
                prim_type, value_expr
            )?;
        }

        ast::TypeName::Unit => {
            write!(out, "{{}}")?;
        }

        ast::TypeName::StrReference(..) => {
            write!(
                out,
                "{}",
                display::iife(|mut f| {
                    writeln!(
                        f,
                        "const [ptr, len] = new Uint32Array(wasm.memory.buffer, {}, 2);",
                        value_expr
                    )?;
                    writeln!(f, "return diplomatRuntime.readString(wasm, ptr, len);")
                })
            )?;
        }
        _ => todo!(),
    }

    Ok(())
}
