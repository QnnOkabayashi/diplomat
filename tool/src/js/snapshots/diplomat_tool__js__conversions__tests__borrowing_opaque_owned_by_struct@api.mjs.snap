---
source: tool/src/js/conversions.rs
expression: out_texts.get(out).unwrap()
---
import wasm from "./wasm.mjs"
import * as diplomatRuntime from "./diplomat-runtime.mjs"
const diplomat_alloc_destroy_registry = new FinalizationRegistry(obj => {
  wasm.diplomat_free(obj["ptr"], obj["size"], obj["align"]);
});

const BorrowingOpaque_box_destroy_registry = new FinalizationRegistry(underlying => {
  wasm.BorrowingOpaque_destroy(underlying);
});

export class BorrowingOpaque {
  constructor(underlying, edges, owned) {
    this.underlying = underlying;
    this.__edges_lifetime_guard = edges;
    if (owned) {
      BorrowingOpaque_box_destroy_registry.register(this, underlying);
    }
  }
}

export class BorrowingStruct {
  constructor(underlying, edges_a) {
    this.opaque = new BorrowingOpaque(diplomatRuntime.ptrRead(wasm, underlying), [...edges_a], true);
    this.x = (new Uint8Array(wasm.memory.buffer, underlying + 4, 1))[0];
  }

  static new(arg_opaque) {
    return (() => {
      const diplomat_receive_buffer = wasm.diplomat_alloc(5, 4);
      wasm.BorrowingStruct_new(diplomat_receive_buffer, arg_opaque.underlying);
      const out = new BorrowingStruct(diplomat_receive_buffer, [arg_opaque]);
      wasm.diplomat_free(diplomat_receive_buffer, 5, 4);
      return out;
    })();
  }
}

